File: src/main/java/dev/wiisportsresorts/duelsplugin/util/Timeout.java
package dev.wiisportsresorts.duelsplugin.util;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;

public class Timeout {
  public Runnable callback;
  Timeout(int delayMs, Runnable callback) {
    this.callback = callback;
    CompletableFuture.delayedExecutor(delayMs, TimeUnit.MILLISECONDS).execute(this::run);
  }

  private void run() {
    callback.run();
  }

  public static Timeout setTimeout(int delayMs, final Runnable callback) {
     return new Timeout(delayMs, callback);
  }
}

File: src/main/kotlin/dev/wiisportsresorts/duelsplugin/util/Timeout.kt
package dev.wiisportsresorts.duelsplugin.util

import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executors
import java.util.concurrent.Future

class Timeout internal constructor(delayMs: Int, var callback: () -> Unit) {
    private var future: Future<Void> = CompletableFuture()

    fun cancel() {
        future.cancel(true)
    }

    private fun run() {
        callback()
    }

    companion object {
        fun setTimeout(delayMs: Int, callback: () -> Unit): Timeout {
            return Timeout(delayMs, callback)
        }
    }

    init {
        Executors.newCachedThreadPool().submit<Any?> {
            Thread.sleep(delayMs.toLong())
            run()
            null
        }
    }
}